File: fixtures/scale/java/large/src/cryptobench/ciphers/scytale/ScytaleCipher.java
package cryptobench.ciphers.scytale;

import cryptobench.core.Cipher;
import cryptobench.util.Alphabet;

/** Wraps text around a rod of fixed circumference and reads down the rod. */
public class ScytaleCipher implements Cipher {
    private final int rows;

    public ScytaleCipher(ScytaleKey key) { this.rows = key.getRows(); }

    @Override
    public String name() { return "scytale"; }

    @Override
    public String encrypt(String plaintext) {
        StringBuilder padded = new StringBuilder(Alphabet.clean(plaintext));
        while (padded.length() % rows != 0) padded.append('X');
        String text = padded.toString();
        int cols = text.length() / rows;
        StringBuilder sb = new StringBuilder();
        for (int c = 0; c < cols; c++) {
            for (int r = 0; r < rows; r++) sb.append(text.charAt(r * cols + c));
        }
        return sb.toString();
    }

    @Override
    public String decrypt(String ciphertext) {
        String text = Alphabet.clean(ciphertext);
        int cols = text.length() / rows;
        StringBuilder sb = new StringBuilder();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) sb.append(text.charAt(c * rows + r));
        }
        return sb.toString();
    }
}

File: fixtures/scale/dart/large/lib/ciphers/scytale/scytale_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'scytale_key.dart';

/// Wraps text around a rod of fixed circumference and reads down the rod.
class ScytaleCipher implements Cipher {
  final int rows;

  ScytaleCipher(ScytaleKey key) : rows = key.rows;

  @override
  String name() => 'scytale';

  @override
  String encrypt(String plaintext) {
    final padded = StringBuffer(cleanAlphabet(plaintext));
    while (padded.length % rows != 0) padded.write('X');
    final text = padded.toString();
    final cols = text.length ~/ rows;
    final sb = StringBuffer();
    for (var c = 0; c < cols; c++) {
      for (var r = 0; r < rows; r++) sb.write(text[r * cols + c]);
    }
    return sb.toString();
  }

  @override
  String decrypt(String ciphertext) {
    final text = cleanAlphabet(ciphertext);
    final cols = text.length ~/ rows;
    final sb = StringBuffer();
    for (var r = 0; r < rows; r++) {
      for (var c = 0; c < cols; c++) sb.write(text[c * rows + r]);
    }
    return sb.toString();
  }
}
